### DISCUSSION TITLE
How to deploy to Posit Connect

### DISCUSSION THREAD

**Discussion by zilch42** [zauberzeug/nicegui#1787] [Q&A] [answered]
### Question

Hi there,

I'm trying to test whether NiceGUI will work on Posit Connect (I can't find any mention of the two together) before I get into learning it proper. Posit Connect works with FastAPI so I think it should work, but I'm currently stuck. 

I'm just trying to upload the basic demo:
```python
# app.py
from nicegui import ui

ui.label('Hello NiceGUI!')

ui.run()
```

I'm deploying it with:
`rsconnect deploy fastapi -n rconnect --entrypoint app .`

The deployment works but I get the following error on the Posit Connect logs when the app tries to start 
```
2023/10/11 22:55:43.243207611 ERROR:    Traceback (most recent call last):
2023/10/11 22:55:43.243245111   File "/opt/rstudio-connect/mnt/app/python/env/lib/python3.9/site-packages/starlette/routing.py", line 677, in lifespan
2023/10/11 22:55:43.243275296     async with self.lifespan_context(app) as maybe_state:
2023/10/11 22:55:43.243276348   File "/opt/rstudio-connect/mnt/app/python/env/lib/python3.9/site-packages/starlette/routing.py", line 566, in __aenter__
2023/10/11 22:55:43.243296243     await self._router.startup()
2023/10/11 22:55:43.243306350   File "/opt/rstudio-connect/mnt/app/python/env/lib/python3.9/site-packages/starlette/routing.py", line 656, in startup
2023/10/11 22:55:43.243311926     handler()
2023/10/11 22:55:43.243312770   File "/opt/rstudio-connect/mnt/app/python/env/lib/python3.9/site-packages/nicegui/nicegui.py", line 75, in handle_startup
2023/10/11 22:55:43.243317822     raise RuntimeError('\n\n'
2023/10/11 22:55:43.243318250 RuntimeError: 
2023/10/11 22:55:43.243322835 
2023/10/11 22:55:43.243323595 You must call ui.run() to start the server.
2023/10/11 22:55:43.243330081 If ui.run() is behind a main guard
2023/10/11 22:55:43.243330511    if __name__ == "__main__":
2023/10/11 22:55:43.243334291 remove the guard or replace it with
2023/10/11 22:55:43.243334726    if __name__ in {"__main__", "__mp_main__"}:
2023/10/11 22:55:43.243338471 to allow for multiprocessing.
2023/10/11 22:55:43.243338936 
```

I've also tried
```python
# app.py
from nicegui import ui

ui.label('Hello NiceGUI!')

app = ui.run()
```

and deploying it with:
`rsconnect deploy fastapi -n rconnect --entrypoint app:app .`

[https://docs.posit.co/connect/user/fastapi/](https://docs.posit.co/connect/user/fastapi/)

Any help would be greatly appreciated

**Comment by rodja** [comment, marked as answer]
Hmm... does Posit Connect also allow arbitrary Python code to be deployed. There is much going on in `ui.run()` which would not be covered if you just specify the entrypoint "app". If that is not possible you can have a look at our FastAPI example: https://github.com/zauberzeug/nicegui/blob/main/examples/fastapi/main.py.

**Reply by zilch42** [reply]
Thanks @rodja for the quick response. The FastAPI example actually worked on Posit Connect (I was a little surprised. I wasn't sure if it would know to run `uvicorn main:app --reload` but it seemed to know what to do with it. I did still have to navigate to `/show` manually to get to the UI but that's probably workable. 

I'm not really familiar with FastAPI either. I was wondering if you could please give me a brief description of what's going on in the example? Do I have a FastAPI running with a GUI sitting on top of it? Does that FastAPI matter or can I just happily ignore it's there and build out my app in `frontend.py`? What would someone normally use a FastAPI with a NiceGUI on top of it for?

Thanks

**Reply by rodja** [reply]
Yes, NiceGUI is just a FastAPI app. The FastAPI example demonstrates how to extend an existing FastAPI App to provide UI through the /show route. A simplified version which serves also the / as NiceGUI page would look like this:

```py
from fastapi import FastAPI
from nicegui import app, ui

fastapi_app = FastAPI()

@ui.page('/')
def show():
    ui.label('Hello, FastAPI!')
    # NOTE dark mode will be persistent for each user across tabs and server restarts
    ui.dark_mode().bind_value(app.storage.user, 'dark_mode')
    ui.checkbox('dark mode').bind_value(app.storage.user, 'dark_mode')

ui.run_with(fastapi_app, storage_secret='secret')

if __name__ == '__main__':
    print('Please start with "uvicorn demo:fastapi_app --reload" (if your file is named "demo.py")')
```

**Reply by zilch42** [reply]
Thank you so much
